### PR TITLE
feat: Extensibility of Verifiable Credential data model

### PR DESCRIPTION
**Pull request by kdimak**
- Provides a mechanism for extension of model by custom data structures which embedded verifiable.Credential

#171 

Signed-off-by: Dima <dkinoshenko@gmail.com>

### CODE REVIEW

File: pkg/doc/verifiable/credential.go [hunk 1]
```diff
@@ -319,7 +365,7 @@ func validate(data []byte, schema *CredentialSchema, opts *credentialOpts) error
 	loader := gojsonschema.NewStringLoader(string(data))
 	result, err := gojsonschema.Validate(schemaLoader, loader)
 	if err != nil {
-		return fmt.Errorf("Validation of Verifiable Credential failed: %w", err)
```
Comment (troyronda) [inline pkg/doc/verifiable/credential.go]: Why %v instead of %w? Is wrapping the error not appropriate here?
(and the other similar changes in this PR).

**Comment by llorllale** [comment]
@kdimak please include a reference to the relevant issue in the PR's description

File: pkg/doc/verifiable/credential.go (hunk 1 above)
Comment (kdimak) [reply pkg/doc/verifiable/credential.go]: agree - I will revert it!

**Comment by kdimak** [comment]
> @kdimak please include a reference to the relevant issue in the PR's description

done!

File: pkg/doc/verifiable/credential.go (hunk 1 above)
Comment (kdimak) [reply pkg/doc/verifiable/credential.go]: Fixed

File: pkg/doc/verifiable/credential.go [hunk 2]
```diff
@@ -239,48 +244,89 @@ func WithDisabledCustomSchemaCheck() CredentialOpt {
...
+
+	credential.Issuer = Issuer{ID: issuerID, Name: issuerName}
+	return
+}
+
+// CredentialBuilder provides a way to extend Credential Data model.
+// It defines a factory method to create new Credential instance and
+// provides decoders needed to fill custom fields of Credential.
+type CredentialBuilder interface {
+	New() *Credential
```
Comment (troyronda) [inline pkg/doc/verifiable/credential.go]: I think it is confusing to have a New() inside an interface.

Perhaps CreateCredential?

File: pkg/doc/verifiable/credential.go [hunk 3]
```diff
@@ -239,48 +244,89 @@ func WithDisabledCustomSchemaCheck() CredentialOpt {
...
+	Decoders() []CredentialDecoder
```
Comment (troyronda) [inline pkg/doc/verifiable/credential.go]: CredentialDecoders?

File: pkg/doc/verifiable/credential.go [hunk 4]
```diff
@@ -239,48 +244,89 @@ func WithDisabledCustomSchemaCheck() CredentialOpt {
...
+	return func(opts *credentialOpts) {
+		opts.cb = cb
+	}
+}
+
+func issuerMapping(data []byte, credential *Credential) (err error) {
+	issuerID, issuerName, err := issuerFromBytes(data)
+	if err != nil {
+		err = fmt.Errorf("JSON unmarshalling of Verifiable Credential bytes failed: %w", err)
+		return
```
Comment (rolsonquadras) [inline pkg/doc/verifiable/credential.go]: Better to have a explicit return than naked return -  [Named Result Parameters](https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters)

File: pkg/doc/verifiable/credential.go [hunk 5]
```diff
@@ -239,48 +244,89 @@ func WithDisabledCustomSchemaCheck() CredentialOpt {
...
+// CredentialBuilder provides a way to extend Credential Data model.
+// It defines a factory method to create new Credential instance and
+// provides decoders needed to fill custom fields of Credential.
+type CredentialBuilder interface {
+	New() *Credential
+	Decoders() []CredentialDecoder
+}
+
+// New Factory method of base Credential.
+func (vc *Credential) New() *Credential {
```
Comment (troyronda) [inline pkg/doc/verifiable/credential.go]: Credential.New returning a Credential seems pretty confusing.

File: pkg/doc/verifiable/credential.go (hunk 4 above)
Comment (llorllale) [reply pkg/doc/verifiable/credential.go]: I opened #216

File: pkg/doc/verifiable/example_credential.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
+/*
```
Comment (llorllale) [inline pkg/doc/verifiable/example_credential.go]: This should be a `_test` file (ie. not in production code)

File: pkg/doc/verifiable/example_credential.go [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+//UniversityDegreeCredential University Degree credential, from examples of https://w3c.github.io/vc-data-model
+type UniversityDegreeCredential struct {
+	c Credential
+
+	Subject *UniversityDegreeSubject
+}
+
+type rawDegree struct {
```
Comment (llorllale) [inline pkg/doc/verifiable/example_credential.go]: I think we would benefit from removing these `raw*` structs altogether from this test and have the non-raw structs hold the json tags. The reason is that I wouldn't want to give the wrong impression and have users think they necessarily have to follow this roundabout way of unmarshalling the credential when a simpler option is available.

File: pkg/doc/verifiable/credential.go (hunk 2 above)
Comment (llorllale) [reply pkg/doc/verifiable/credential.go]: And its method set seems disjointed. I suggest splitting them into two new options, perhaps with names `WithTemplate` and `WithDecoder`.

File: pkg/doc/verifiable/credential.go (hunk 4 above)
Comment (kdimak) [reply pkg/doc/verifiable/credential.go]: Useful link regarding code review comments! Thanks, I will update this.

File: pkg/doc/verifiable/example_credential.go (hunk 6 above)
Comment (kdimak) [reply pkg/doc/verifiable/example_credential.go]: OK!

File: pkg/doc/verifiable/example_credential.go (hunk 7 above)
Comment (kdimak) [reply pkg/doc/verifiable/example_credential.go]: OK!

File: pkg/doc/verifiable/credential.go (hunk 2 above)
Comment (kdimak) [reply pkg/doc/verifiable/credential.go]: I like the simplification @llorllale proposed!

File: pkg/doc/verifiable/credential.go (hunk 5 above)
Comment (kdimak) [reply pkg/doc/verifiable/credential.go]: agree!

File: pkg/doc/verifiable/example_credential_test.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+func (udc *UniversityDegreeCredential) decode(dataJSON []byte, credential *Credential) error {
+	return json.Unmarshal(dataJSON, udc)
+}
+
+func TestCredentialExtensibility(t *testing.T) {
+	udc := &UniversityDegreeCredential{}
+	cred, err := NewCredential(
+		[]byte(validCredential),
```
Comment (troyronda) [inline pkg/doc/verifiable/example_credential_test.go]: We should add a utility function that would supply the credential type from the marshalled data. From there, someone could determine which type to instantiate.

Comment (troyronda) [reply pkg/doc/verifiable/example_credential_test.go]: Created https://github.com/hyperledger/aries-framework-go/issues/218

**Pull request merged**